packages/paymasters/contracts/PermitERC20UniswapV3Paymaster.sol

```diff
@@ -147,33 +225,109 @@ contract PermitERC20UniswapV3Paymaster is BasePaymaster, ERC2771Recipient {
     internal
     override
     {
-        (address payer,) = abi.decode(context, (address, uint256));
-        uint256 ethActualCharge = relayHub.calculateCharge(gasUseWithoutPost + gasUsedByPost, relayData);
-        _depositProceedsToHub(ethActualCharge);
-        uint256 remainingTokenBalance = token.balanceOf(address(this));
-        require(token.transfer(payer, remainingTokenBalance), "failed refund");
-        emit TokensCharged(gasUseWithoutPost, gasUsedByPost, ethActualCharge);
-    }
-
-    function _depositProceedsToHub(uint256 ethActualCharge) private {
-        UniswapV3Helper.swapToEth(
-            address(weth),
-            address(token),
-            ethActualCharge,
-            uniswapPoolFee,
-            uniswap
-        );
-        relayHub.depositFor{value : address(this).balance}(address(this));
+        (address payer, uint256 priceQuote, uint256 tokenPreCharge) = abi.decode(context, (address, uint256, uint256));
+        IERC20Metadata token = _getTokenFromPaymasterData(relayData.paymasterData);
+        (uint256 tokenActualCharge, uint256 ethActualCharge) = _calculateCharge(relayData, gasUseWithoutPost + gasUsedByPost, priceQuote, tokensSwapData[token].priceDivisor);
+        require(tokenActualCharge <= tokenPreCharge, "actual charge higher");
+        token.safeTransfer(payer, tokenPreCharge - tokenActualCharge);
+
+        emit TokensCharged(gasUseWithoutPost, gasUsedByPost, tokenActualCharge, ethActualCharge);
+        _refillHubDepositIfNeeded(ethActualCharge);
+        _withdrawToOwnerIfNeeded();
+    }
+
+    function _refillHubDepositIfNeeded(uint256 ethActualCharge) private {
+        uint256 hubBalance = relayHub.balanceOf(address(this));
+        if (hubBalance >= minHubBalance + ethActualCharge) {
+            return;
+        }
+        uint256 depositAmount = targetHubBalance - hubBalance + ethActualCharge;
+        _refillHubDeposit(depositAmount);
+    }
+
+    function _refillHubDeposit(uint256 depositAmount) private {
+        uint256 balance = address(this).balance;
+        uint256 amountSwapped = 0;
+        if (balance < depositAmount) {
+            for (uint256 i = 0; i < tokens.length && balance + amountSwapped < depositAmount; i++) {
```

Comment:
Also a bit concerned about the potentially very, very expensive loop in the `postRelayCall`. 
It can, quite possibly, make the Paymaster inoperable as it may require waaay more gas then available here.
I think for the "forced" swap in users' transaction we limit to a single swap or something like that.

Reply:
Added a threshold that pretty much solves it